File: typescript/src/tracing.ts
import { context, propagation, type Context } from "@opentelemetry/api";
import { CompositePropagator, W3CBaggagePropagator, W3CTraceContextPropagator } from "@opentelemetry/core";
import { AsyncLocalStorageContextManager } from "@opentelemetry/context-async-hooks";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { BatchSpanProcessor, SimpleSpanProcessor, type SpanExporter } from "@opentelemetry/sdk-trace-base";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { headersGetter } from "./logging/_headers";

export type { Context };

export const DEFAULT_ENDPOINT = "http://tempo:4318";

export interface SetupTracingOptions {
  service: string;
  endpoint?: string;
  /** Réservé aux tests : branche un exporteur (ex. `InMemorySpanExporter`) sur
   * un `SimpleSpanProcessor` (export synchrone à `span.end()`) plutôt que sur
   * le `BatchSpanProcessor` du fonctionnement normal. */
  exporter?: SpanExporter;
  /** Désactive l'auto-enregistrement du handler SIGTERM interne (défaut
   * `true`). À mettre à `false` si le consommateur orchestre déjà son propre
   * arrêt propre (ex. drainage de boucles de fond) — il doit alors appeler
   * `shutdownTracing()` lui-même, dans sa propre séquence, plutôt que de
   * laisser deux handlers SIGTERM indépendants se disputer l'ordre d'arrêt. */
  handleShutdownSignal?: boolean;
}

/** Capture le contexte de trace actif (le span courant, posé par
 * `withRequestLogging` ou un appel englobant à `withTraceContext`), pour le
 * réinjecter plus tard via `withTraceContext` — utile quand un chemin de code
 * (ex. une Server Action Next.js, jamais enveloppée par `withRequestLogging`)
 * doit capturer le contexte tôt puis le réappliquer explicitement autour
 * d'appels sortants plus loin dans l'exécution. */
export function activeTraceContext(): Context {
  return context.active();
}

/** Extrait le contexte de trace d'un `traceparent` entrant (ex. `headers()`
 * d'une Server Action Next.js, qui ne passe par aucun middleware). */
export function extractTraceContext(headers: Headers): Context {
  return propagation.extract(context.active(), headers, headersGetter);
}

/** Réapplique explicitement un contexte capturé plus tôt (voir
 * `activeTraceContext`/`extractTraceContext`) autour de `fn` — équivalent
 * OpenTelemetry de l'ancien `runWithTraceId`. */
export function withTraceContext<T>(ctx: Context, fn: () => T): T {
  return context.with(ctx, fn);
}

let provider: NodeTracerProvider | undefined;

// Propagateur composite W3C explicite : `NodeTracerProvider.register()` ne
// pose PAS de propagateur par défaut (contrairement à l'API OTel Python, qui
// en installe un depuis `OTEL_PROPAGATORS` à l'import), il faut le fournir.
const propagator = new CompositePropagator({
  propagators: [new W3CTraceContextPropagator(), new W3CBaggagePropagator()],
});

export function setupTracing(options: SetupTracingOptions): void {
  const spanProcessor = options.exporter
    ? new SimpleSpanProcessor(options.exporter)
    : new BatchSpanProcessor(
        new OTLPTraceExporter({
          url: `${(options.endpoint ?? process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? DEFAULT_ENDPOINT).replace(/\/$/, "")}/v1/traces`,
        }),
      );

  provider = new NodeTracerProvider({
    resource: resourceFromAttributes({ "service.name": options.service }),
    spanProcessors: [spanProcessor],
  });

  provider.register({
    contextManager: new AsyncLocalStorageContextManager(),
    propagator,
  });

  // Flush des spans bufferisés par le BatchSpanProcessor avant l'arrêt du
  // process — sans ça les spans en attente sont perdus sur un redéploiement
  // Swarm (SIGTERM). Next.js n'a pas de hook de lifecycle équivalent au
  // lifespan FastAPI / on_cleanup aiohttp côté Python, d'où ce handler
  // process-level par défaut. Si le consommateur a déjà sa propre séquence
  // d'arrêt (ex. drainage de boucles de fond), `handleShutdownSignal: false`
  // désactive ce handler pour éviter deux listeners SIGTERM indépendants qui
  // se disputent l'ordre d'arrêt — le consommateur appelle alors
  // `shutdownTracing()` lui-même, au bon endroit de sa propre séquence.
  if (options.handleShutdownSignal ?? true) {
    process.once("SIGTERM", () => {
      void shutdownTracing().finally(() => process.exit(0));
    });
  }
}

export async function shutdownTracing(): Promise<void> {
  if (provider) {
    await provider.shutdown();
    provider = undefined;
  }
}

File: python/src/chauff_cmn/tracing.py
"""Configuration du tracing distribué OpenTelemetry, exporté en OTLP vers
Tempo. `setup_tracing()` est à appeler une fois au démarrage du service ;
les middlewares/clients de `chauff_cmn.logging` créent ensuite de vrais spans
sur ce provider (récupéré via `opentelemetry.trace.get_tracer()`, safe à
appeler avant `setup_tracing()` grâce au proxy tracer de l'API OTel).
"""

import atexit
import os
from typing import Optional

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, SimpleSpanProcessor, SpanExporter

__all__ = ["setup_tracing", "shutdown_tracing"]

DEFAULT_ENDPOINT = "http://tempo:4318"

_provider: Optional[TracerProvider] = None


def setup_tracing(service: str, endpoint: Optional[str] = None, exporter: Optional[SpanExporter] = None) -> None:
    """Installe un `TracerProvider` global qui exporte en OTLP/HTTP vers Tempo.

    `endpoint` (base URL, sans `/v1/traces`) retombe sur la variable d'env
    standard `OTEL_EXPORTER_OTLP_ENDPOINT`, puis sur `http://tempo:4318`
    (hostname Swarm interne, cf. `deploy/stacks/core/tempo.yml`).

    `exporter` est réservé aux tests : un exporteur fourni explicitement (ex.
    `InMemorySpanExporter`) est branché sur un `SimpleSpanProcessor` (export
    synchrone à `span.end()`) plutôt que sur le `BatchSpanProcessor` utilisé en
    fonctionnement normal, pour que les spans soient immédiatement visibles
    sans attendre un flush.
    """
    global _provider

    provider = TracerProvider(resource=Resource.create({"service.name": service}))

    if exporter is None:
        base = (endpoint or os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT") or DEFAULT_ENDPOINT).rstrip("/")
        provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{base}/v1/traces")))
    else:
        provider.add_span_processor(SimpleSpanProcessor(exporter))

    trace.set_tracer_provider(provider)
    _provider = provider
    atexit.register(shutdown_tracing)


def shutdown_tracing() -> None:
    """Flush et arrête le `TracerProvider` — les spans bufferisés par le
    `BatchSpanProcessor` sont perdus si le process se termine sans ça.
    `atexit` (posé par `setup_tracing`) ne suffit pas sur un `SIGTERM` non
    intercepté : à brancher explicitement sur le hook d'arrêt du framework
    (lifespan FastAPI, `on_cleanup` aiohttp, etc.).
    """
    global _provider
    if _provider is not None:
        _provider.shutdown()
        _provider = None
